### DISCUSSION TITLE
Undocumented P-Code instructions

### DISCUSSION THREAD

**Discussion by niconaus** [NationalSecurityAgency/ghidra#4130] [Q&A]
I was digging through the API documentation, and came across several P-Code instructions that are not listed in the P-Code language documentation. Can anyone explain what the following instructions do, and how they are used (number of args, etc)?

EXTRACT
INSERT
PCODE_MAX
SEGMENTOP
UNIMPLEMENTED

**Comment by GhidorahRex** [comment]
I can't help with the exact number of parameters (it's not my area of expertise), but I can at least offer an explanation of what these are all doing:

Extract and Insert are used for extracting/inserting bits into a varnode via the bracket notation.
Extract:
foo = bar[2,3]
Insert:
foo[2,3] = bar

pcode_max doesn't do anything. It's a placeholder in the table to say that we've encountered an unsupported pcodeop (and  gone off the end of the list of pcode ops).

segmentop is used for memory mappings. It basically takes a memory segment pointer and an address within that segment and joins them together to create an address.

unimplemented is a placeholder to identify that a particular assembly instruction is unimplemented.

**Comment by ghidracadabra** [comment]
INSERT and EXTRACT are described in the file `additionalpcode.html`, which is located in the `docs/languages/html` directory of a Ghidra distribution.

**Comment by urn1ce** [comment]
**SEGMENTOP** still not described in `docs/languages/html`, wish to fix it.
